internal/charmstore/migrations_test.go

```diff
@@ -793,6 +848,600 @@ func (s *migrationsSuite) TestPopulateWriteACLSomeUpdates(c *gc.C) {
+		},
+	}, entity2...)
+	s.insertEntity(c, &mongodoc.Entity{
+		URL:      id3,
+		BlobHash: "django-1",
+		ExtraInfo: map[string][]byte{
+			"bzr-owner": []byte(`"who"`),
+		},
+	}, entity2...)
+	s.insertEntity(c, &mongodoc.Entity{
+		URL:      id4,
+		BlobHash: "django-1",
+		ExtraInfo: map[string][]byte{
+			"bzr-owner": []byte(`"dalek"`),
+		},
+	}, entity2...)
+	s.insertEntity(c, &mongodoc.Entity{
+		URL:      id5,
+		BlobHash: "rails-1",
+		ExtraInfo: map[string][]byte{
+			"bzr-owner": []byte(`"ace"`),
+		},
+	}, entity2...)
+	s.insertEntity(c, &mongodoc.Entity{
+		URL:      id6,
+		BlobHash: "django-2",
+		ExtraInfo: map[string][]byte{
+			"bzr-owner": []byte(`"who"`),
+		},
+		PromulgatedURL: id7,
+	}, entity3...)
+
+	s.insertBaseEntity(c, &mongodoc.BaseEntity{URL: baseURL(id1)}, baseEntity2...)
+	s.insertBaseEntity(c, &mongodoc.BaseEntity{URL: baseURL(id2)}, baseEntity2...)
+	s.insertBaseEntity(c, &mongodoc.BaseEntity{
+		URL:         baseURL(id3),
+		Promulgated: mongodoc.True,
+	}, baseEntity3...)
+	s.insertBaseEntity(c, &mongodoc.BaseEntity{URL: baseURL(id4)}, baseEntity2...)
+	s.insertBaseEntity(c, &mongodoc.BaseEntity{URL: baseURL(id5)}, baseEntity2...)
+
+	// Start the server.
+	err := s.newServer(c)
+	c.Assert(err, gc.IsNil)
+
+	// Ensure entities have been updated correctly.
+	s.checkCount(c, s.db.Entities(), 5)
+	s.checkEntity(c, &mongodoc.Entity{
+		URL:                 id2,
+		BaseURL:             baseURL(id2),
+		BlobHash:            "rails-1",
+		User:                id2.User,
+		Name:                id2.Name,
+		Revision:            id2.Revision,
+		Series:              id2.Series,
+		PromulgatedRevision: -1,
+		ExtraInfo: map[string][]byte{
+			"bzr-owner": []byte(`"who"`),
+		},
+	})
+	s.checkEntity(c, &mongodoc.Entity{
+		URL:                 id3,
+		BaseURL:             baseURL(id3),
+		BlobHash:            "django-1",
+		User:                id3.User,
+		Name:                id3.Name,
+		Revision:            id3.Revision,
+		Series:              id3.Series,
+		PromulgatedURL:      id1,
+		PromulgatedRevision: id1.Revision,
+		ExtraInfo: map[string][]byte{
+			"bzr-owner": []byte(`"who"`),
+		}})
+	s.checkEntity(c, &mongodoc.Entity{
+		URL:                 id4,
+		BaseURL:             baseURL(id4),
+		BlobHash:            "django-1",
+		User:                id4.User,
+		Name:                id4.Name,
+		Revision:            id4.Revision,
+		Series:              id4.Series,
+		PromulgatedRevision: -1,
+		ExtraInfo: map[string][]byte{
+			"bzr-owner": []byte(`"dalek"`),
+		},
+	})
+	s.checkEntity(c, &mongodoc.Entity{
+		URL:                 id5,
+		BaseURL:             baseURL(id5),
+		BlobHash:            "rails-1",
+		User:                id5.User,
+		Name:                id5.Name,
+		Revision:            id5.Revision,
+		Series:              id5.Series,
+		PromulgatedRevision: -1,
+		ExtraInfo: map[string][]byte{
+			"bzr-owner": []byte(`"ace"`),
+		},
+	})
+	s.checkEntity(c, &mongodoc.Entity{
```

Comment:
If, instead of putting the auto-field-fill logic inside insertEntity, we put it into
a makeEntity function, then these places could benefit from it too.

e.g.
   func makeEntity(e mongodoc.Entity) *mongodoc.Entity


    s.checkEntity(c, makeEntity(mongodoc.Entity{
        URL: id6,
        BlobHash: "django-2", 
        PromulgatedURL: id7,
        ExtraInfo: map[string][]byte{ ...},
   }))